### ISSUE TITLE
Use incoming tx weight per second to predict the future

### ISSUE THREAD

**Issue by softsimon** [mempool/mempool#65]
Transactions ETA are now estimated based on blocks in queue * 10, which doesn't give an accurate time prediction due to constantly incoming transactions.

Make an algorithm that takes tx weight/second and their fees into account to predict the state of the future mempool for both transaction ETA and fee estimation.

**Comment by wiz** [comment]
As a temporary fix to improve fee estimation, how about we just average the low + medium + high estimates together for a new blended estimate rate?

**Comment by wiz** [comment]
@natsoni can you take a look at this?

**Comment by orangesurf** [comment]
# Block confirmation time estimator

Given a Bitcoin transaction at a known position in the mempool, estimate how long until it receives its first confirmation, accounting for the probabilistic nature of block arrival and competition from higher-fee transactions.

## Parameters

| Variable | Description | Default |
|----------|-------------|---------|
| `block_interval` | Mean time between blocks (minutes) | 10 |
| `block_position` | How many blocks deep your transaction is in the mempool (e.g. 3 means three blocks must be mined for your tx to be included) | 3 |
| `confidence` | Probability threshold (0–1) | 0.95 |
| `incoming_rate` | Rate of new transactions entering the mempool ahead of yours (vB/s) | 500 |
| `max_block_rate` | Maximum block capacity as a rate: 1,000,000 vB / 600s | 1666.6 |

## Simple model

Blocks arrive as a Poisson process. If your transaction sits in the 3rd mempool block, three blocks need to be mined before it confirms. We want the shortest wait time where we're confident enough blocks have arrived.

This assumes every block fully advances your transaction's position. In practice new transactions paying higher fees arrive between blocks and consume capacity that would otherwise clear the queue ahead of you.

## Block discount factor

Each block has ~1,000,000 vB of capacity. During the `block_interval` between blocks, new higher-fee transactions consume `incoming_rate × block_interval × 60` vB of that space. The fraction of each block's capacity that actually advances your position is:

$$\text{discount} = 1 - \frac{\text{incomingRate}}{\text{maxBlockRate}}$$

When `discount = 1`, no new transactions enter ahead of you and each block brings you 1MvB closer to confirmation. When `discount → 0`, new transactions fill blocks as fast as they are produced and you make no progress.

This stretches the apparent block interval:

$$\text{effectiveInterval} = \frac{\text{blockInterval}}{\text{discount}}$$

## Finding the minimum wait time

We need the smallest number of minutes `t` where we're confident that `block_position` blocks have arrived, using `effective_interval` as the adjusted block rate.

**As pseudocode:**

```
for t = 1, 2, 3, ...:
    if poisson_cdf(block_position - 1, t / effective_interval) ≤ (1 - confidence):
        return t
```

Keep increasing `t` until the probability of getting fewer than `block_position` blocks drops below your error tolerance.

**In plain English:** Find the smallest whole number of minutes where a Poisson distribution with rate `t / effective_interval` gives you at least a 95% chance of `block_position` or more arrivals.

 **As a reference table** (`block_interval = 10`, `confidence = 0.95`):

  | block_position | incoming_rate (vB/s) | discount | effective_interval (min) | min_wait (min) |
  |:--------------:|---------------------:|:--------:|:------------------------:|:--------------:|
  | 1              |                    0 |     1.00 |                     10.0 |            ~30 |
  | 1              |                  500 |     0.70 |                     14.3 |            ~43 |
  | 3              |                    0 |     1.00 |                     10.0 |            ~63 |
  | 3              |                  500 |     0.70 |                     14.3 |            ~90 |
  | 6              |                    0 |     1.00 |                     10.0 |           ~106 |
  | 6              |                  500 |     0.70 |                     14.3 |           ~151 |
  | 6              |                 1000 |     0.40 |                     25.0 |           ~263 |

  And the corrected worked example:

  ## Worked example

  With `block_interval = 10`, `block_position = 6`, `confidence = 0.95`, `incoming_rate = 500`:

  $$\text{discount} = 1 - \frac{500}{1666.6} = 0.7$$

  $$\text{effectiveInterval} = \frac{10}{0.7} \approx 14.3 \text{ min}$$

  Mean wait is $6 \times 14.3 \approx 86$ min, but for 95% confidence we need to account for the Poisson tail. Iterating over integer minutes, `min_wait ≈ 151` min.

  Without competition (`incoming_rate = 0`), the same calculation gives ~106 min — the extra ~45 minutes is the cost of 500 vB/s of fee competition.
